### PR TITLE
New interface

### PR DESCRIPTION
**Pull request by andrewmoise**
This is an implementation of the new simpler interface:

* browser()
* browser_string()
* browser_version() (+ _major(), _minor(), _beta())
* robot()
* robot_string()
* os()
* os_string()
* os_version() (+ _major(), _minor(), _beta())
* device()
* device_string()
* engine()
* engine_string()
* engine_version() (+ _major(), _minor(), _beta())

  ... so everything is much more consistent. I also implemented new functionality for the parts of that interface that weren't already defined. I tried not to break existing usage of the library in any major way; for example, the interface for _version() is the same as what public_version() used to be, and the output of _string() is the same for all relatively common cases. I also made compatibility functions for things like public_version() and robot_name() where the name of something changed.
  Let me know what you think and anything that you see should be changed or added.

### CODE REVIEW

File: lib/HTTP/BrowserDetect.pm [hunk 1]
```diff
@@ -394,68 +395,96 @@ sub user_agent {
 # Private method -- Set up the basics (browser and misc attributes)
 # for a new user-agent string
 sub _init_core {
```
Comment (oalders) [inline lib/HTTP/BrowserDetect.pm]: I don't think we need to preserve this functionality here.  I'm happy to see it go and force everyone to use ->new()

**Comment by oalders** [comment]
Wow.  Impressive stuff.  Looks good to me.  My only other comment is if you've thought about just returning the entire version for the new *_version() methods.  We could also have a *_full_version method or I guess people could concatenate the beta part themselves.  Would probably be handy if we just made it easy though.

**Comment by andrewmoise** [comment]
Okay, user_agent() with argument is removed. I added a die() if someone tries to call it, so that existing code will fail instead of mysteriously returning wrong stuff.

My thought for the _version() was that there are several equally sensible ways of doing it, so the way we're already doing it (that existing code depends on) wins the tiebreaker. Also, it's slightly convenient that the _version() output can be converted to a float. My thought is that _full_version() isn't necessary (we already have four different version functions for everything we can have a version of), although I guess the other side of that is "we already have four version functions, what's one more?" :-) It's your call.

**Comment by oalders** [comment]
That makes sense to me.  I was mostly just curious as to what you thought about it.  I think version objects at some point would be a good solution and, in the meantime, nobody has asked for the full version, so I think this is good.  I'll probably bump this to 2.0 because of the breaking change with the user_agent() method.

**Comment by oalders** [comment]
Coverage looks really good: http://cpancover.com/latest//HTTP-BrowserDetect-2.00/index.html

**Comment by andrewmoise** [comment]
Excellent, looks good! Version objects would be nice; I do think there's a lot of value to keeping things the same as they have been so everyone's code keeps working though.

2.0 sounds like a good idea, since we revamped the interface and broke some existing behavior. The obvious other interface change I can see making would be to add robot_version().

This is sort of an idea of how I'm using this:

```
    my $indexerflag = $browser_detect->robot();

    my $browser_details;
    if ($browser_detect->robot) {
        $browser_details = $browser_detect->robot_string;
    } elsif ($browser_detect->browser_string) {
        $browser_details = $browser_detect->browser_string;
        if ($browser_detect->public_version) {
            $browser_details .= ' ' . $browser_detect->public_version . $browser_detect->public_beta;
        }
        if ($browser_detect->mobile || $browser_detect->tablet) {
            if ($browser_detect->device_string) {
                $browser_details .= '/' . $browser_detect->device_string;
            }
        } else {
            if ($browser_detect->os_string) {
                $browser_details .= '/' . $browser_detect->os_string;
            }
        }
    }
```

As things stand now, the code does a fairly decent job of identifying and fully describing most things in my logs. The obvious gap still remaining is that there are a lot of robots that it identifies as not robots because it doesn't recognize them. The real test will be when I make it work for one month's logs, and then it successfully parses everything for the *next* month's logs. :-)

I did some more work fixing some outstanding issues and implementing more stuff that makes my log files parse more or less correctly almost all the time; I'll aim to upload that in the near future.

**Comment by oalders** [comment]
@andrewmoise robot_version sounds useful. I hadn't considered that.  Is it really much different from browser_version?

**Comment by andrewmoise** [comment]
There is a case where it's different: there's an odd special case where a robot identifies itself as a particular browser (e.g. googlebot-mobile pretends to be Mobile Safari so that it gets the mobile version of pages). In those cases, we report both a browser and a robot; for example:

`Mozilla/5.0 (iPhone; U; CPU iPhone OS 4_1 like Mac OS X; en-us) AppleWebKit/532.9 (KHTML, like Gecko) Version/4.0.5 Mobile/8B117 Safari/6531.22.7 (compatible; Googlebot-Mobile/2.1; +http://www.google.com/bot.html)`

... is identified as follows:

```
browser: "safari"
browser_version: "4.0"
mobile: 1
robot: "googlemobile"
robot_version: "2.1" # proposed
```

I don't know whether that was intended behavior, but to me it seems right. Because we're doing things that way, users who ask about browser and mobile vs. non mobile status get the answer that's going to make googlebot-mobile function as intended, and users who ask about robot vs. non robot status also get the answer that they're looking for. Therefore I've been preserving that behavior as I've been working, and I explained that case in the help.

One other related issue: It'd be nice to be able to identify a browser as multiple things with a version for each. For example, some Galeon versions are a wrapper around some particular version of Firefox, AOL 9.5 is a wrapper around MSIE 7.0, etc. It'd be nice if we could accurately report both the version of AOL and the version of MSIE that's involved. I'm reluctant to overcomplicate the interface to handle something that's extremely rare, but if there were a simple way to do it it'd be nice to correctly handle that case.

I should make another pull request soon; I've been doing more work intermittently (fixing a few bug reports that didn't have anyone obviously working on them, adding some browsers, and adding detection for more robots). As soon as I have things cleaned up a bit I'll send them along.

-Andrew

**Comment by oalders** [comment]
I see what you're saying.  Yes, that makes good sense to have a `robot_version()`.

As far as the wrappers go, I wonder if we'd just have a method that would return another HTTP::BrowserDetect object so the original object would identify as what the browser is pretending to be and the new object would return info about what the browser actually is.  That doesn't seem incredibly simple either, though.  :)

Thanks for all of the work you've put into this!

**Comment by oalders** [comment]
Hi Andrew,

I've just launched http://browserdetect.org  I'll try to keep it up to date with the versions as they get released.

**Comment by andrewmoise** [comment]
Sounds good. I haven't had as much time to devote to this recently, but I
do have a few changes I still need to polish and push back. I'll aim to
make time for that sometime this weekend.
  -Andrew

**Comment by dougwilson** [comment]
@oalders FWIW, this was the little app I used to use to test changes: https://github.com/dougwilson/perl5-http-browserdetect/commit/8196f58b23746a1d417e5e51cda764206d1217d5 It was useful for me to run locally and access it from devices on my network. You're welcome to incorporate anything useful into that new site :)

**Comment by oalders** [comment]
@dougwilson Thanks for this. I'm new to Mojo, so it's interesting to see how you went about things.  Funny that we had the same general idea.  :)

**Pull request merged**